src/bitmap/backend/atomic_bitmap.rs

```diff
@@ -136,38 +137,35 @@ impl Bitmap for AtomicBitmap {
 
 impl Default for AtomicBitmap {
     fn default() -> Self {
-        AtomicBitmap::new(0, 0x1000)
+        // SAFETY: Safe as `0x1000` is non-zero.
+        AtomicBitmap::new(0, unsafe { NonZeroUsize::new_unchecked(0x1000) })
     }
 }
 
 #[cfg(feature = "backend-mmap")]
 impl NewBitmap for AtomicBitmap {
     fn with_len(len: usize) -> Self {
-        let page_size;
-
         #[cfg(unix)]
-        {
-            // SAFETY: There's no unsafe potential in calling this function.
-            page_size = unsafe { libc::sysconf(libc::_SC_PAGE_SIZE) };
-        }
+        // SAFETY: There's no unsafe potential in calling this function.
+        let page_size = unsafe { libc::sysconf(libc::_SC_PAGE_SIZE) };
 
         #[cfg(windows)]
-        {
+        // SAFETY: It's safe to initialize this object from a zeroed memory region.
+        // It's safe to call `GetSystemInfo` as the pointer is based on the address
+        // of the previously initialized `sysinfo` object.
+        let page_size = unsafe {
             use winapi::um::sysinfoapi::{GetSystemInfo, LPSYSTEM_INFO, SYSTEM_INFO};
-
-            // It's safe to initialize this object from a zeroed memory region.
-            let mut sysinfo: SYSTEM_INFO = unsafe { std::mem::zeroed() };
-
-            // It's safe to call this method as the pointer is based on the address
-            // of the previously initialized `sysinfo` object.
-            unsafe { GetSystemInfo(&mut sysinfo as LPSYSTEM_INFO) };
-
-            page_size = sysinfo.dwPageSize;
-        }
+            let mut sysinfo: SYSTEM_INFO = std::mem::zeroed();
+            GetSystemInfo(&mut sysinfo as LPSYSTEM_INFO);
+            sysinfo.dwPageSize
+        };
 
         // The `unwrap` is safe to use because the above call should always succeed on the
         // supported platforms, and the size of a page will always fit within a `usize`.
-        AtomicBitmap::new(len, usize::try_from(page_size).unwrap())
+        AtomicBitmap::new(
+            len,
+            NonZeroUsize::try_from(usize::try_from(page_size).unwrap()).unwrap(),
```

Comment:
This is not a good sign, we change a divided by zero into panic:)

Reply:
What do you think of returning `Result` or `Option` from this function with `NonZeroUsize::try_from(usize::try_from(page_size).unwrap())?` ?

Is there another approach you would propose?

Reply:
I think this is a good example why one could argue that this may not be worth the effort... Realistically, people will replace a division by zero (where Rust would panic) with accidental unsafety or manual panics. Not sure what is gained there. The only alternative is forcing Result<> into paths that are usually very unlikely to fail, cluttering the code... And all of that at the cost of breaking existing API and forcing people to update their code.

I got no strong opinion on this, but after still waiting for crates to update after the recent vm-memory API breaks, I am a bit worried...

Reply:
I understand this, I am not strongly opinionated on pushing for this change.

Reply:
> This is not a good sign, we change a divided by zero into panic:)
A divide by zero is already a panic in rust, both in release and debug builds.

I also don't think using `Result` or `Option` here is a good idea, for the reasons @Ablu mentions. Passing in zero for page size is fundamentally wrong, so replacing the implicit "divide by zero" panic with one that's a bit more informative (e.g. "attempt to pass zero for page size") or using a type-system level approach as in this PR would be my preferred approach.

Reply:
Isn't updating the AtomicBitmap::new to take as a parameter NonZeroUsize also a breaking change? Why is this better than returning a Result?

Reply:
It is, yeah. I greatly prefer it over the `Result` approach though, because `NonZeroUsize` already prevents you at compile time from passing in nonsense.

Reply:
I am still a bit worried that this is a lot of code churn for very little gain... But I have no strong opinion here.

Are other crates within rust-vmm using the changed public API? Should we prepare some PRs to demonstrate how those need to be adjusted?

Reply:
> I think this is outside the scope of this PR. I don't think any PR that introduces a breaking change should also require PR's to all other rust-vmm crates using breaking change.
> 
> I would suggest if we wanted to move towards this pattern we should move towards a more monorepo-like design to support it.

We brought this up during the last rust-vmm sync since the lack of such investigation caused us to get stuck with some state where we were unable to update vm-memory for a couple of months. The echo there was that we should encourage people to submit PRs that demonstrate the scale of changes that are required on the rest of the rust-vmm ecosystem. IMHO it plays a critical role in deciding the impact of breaking changes and helps with timing the release.

Reply:
I ran `cargo test` on vm-virtio with a `crates-io.patch` directive for vm-memory pointing this PR and everything passed, so I think we are fine here! Maybe that is something we can add to the CI in vm-memory as an optional check?

Reply:
I've created https://github.com/rust-vmm/vm-memory/issues/271 to track this.